submission/races.go

```diff
@@ -0,0 +1,208 @@
+)
+
+const (
+	// PostBatchInterval is duration between parallel batch call and subsequent requests to Logs within group.
+	PostBatchInterval = time.Second
+)
+
+type safeStringSet struct {
+	logURLs map[string]bool
+	mu      sync.Mutex
+}
+
+// Returns whether add was successful: no entry before, entry added.
+func (gList *safeStringSet) add(logURL string) bool {
+	gList.mu.Lock()
+	defer gList.mu.Unlock()
+	if gList.logURLs[logURL] {
+		return false
+	}
+	gList.logURLs[logURL] = true
+	return true
+}
+
+func (gList *safeStringSet) has(logURL string) bool {
+	gList.mu.Lock()
+	defer gList.mu.Unlock()
+	exist := gList.logURLs[logURL]
+	return exist
+}
+
+func (gList *safeStringSet) sumFor(URLs map[string]bool) int {
+	var entries int
+	gList.mu.Lock()
+	defer gList.mu.Unlock()
+	for url := range URLs {
+		if gList.logURLs[url] {
+			entries++
+		}
+	}
+	return entries
+}
+
+type result struct {
+	logURL string
+	sct    *ct.SignedCertificateTimestamp
+	err    error
+}
+
+// postInterval calculates duration for consequent call. For first parallelStart calls duration is 0, while every next one gets additional dur interval.
+func postInterval(idx int, parallelStart int, dur time.Duration) time.Duration {
+	if idx < parallelStart {
+		return time.Duration(0)
+	}
+	return time.Duration(idx+1-parallelStart) * dur
+}
+
+// Submitter is interface wrapping Log-request-response cycle and any processing.
+type Submitter interface {
+	SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error)
+}
+
+func groupRace(ctx context.Context, chain []ct.ASN1Cert, group *ctpolicy.LogGroupInfo, parallelStart int,
+	results chan result, requested *safeStringSet, receivedSCTFrom *safeStringSet, submitter Submitter) {
+	groupURLs := make([]string, len(group.LogURLs))
+	var i int
+	for logURL := range group.LogURLs {
+		groupURLs[i] = logURL
+		i++
+	}
+	for i, urlNum := range rand.Perm(len(groupURLs)) {
+		go func(i int, logURL string) {
+			time.Sleep(postInterval(i, parallelStart, PostBatchInterval))
+			if ctx.Err() != nil {
+				return
+			}
+			if firstRequested := requested.add(logURL); !firstRequested {
+				return
+			}
+			if receivedSCTFrom.sumFor(group.LogURLs) >= group.MinInclusions {
+				return
+			}
+			sct, err := submitter.SubmitToLog(ctx, logURL, chain)
+			if err != nil {
+				if err != context.Canceled {
+					// TODO(Mercurrent) Submission failed. Log it
+				}
+				results <- result{logURL: logURL, err: err}
+				return
+			}
+			// verify SCT
+			results <- result{sct: sct, logURL: logURL}
+		}(i, groupURLs[urlNum])
+	}
+}
+
+func parallelNums(groups ctpolicy.LogPolicyData) map[string]int {
+	nums := make(map[string]int)
+	var subsetSum int
+	for _, g := range groups {
+		if !g.IsBase {
+			nums[g.Name] = g.MinInclusions
+			subsetSum += g.MinInclusions
+		}
+	}
+	nums[ctpolicy.BaseName] -= subsetSum
+	return nums
+}
+
+func totalLogs(groups ctpolicy.LogPolicyData) int {
```

Comment:
While I agree reg. totalLogs (moved), I'm in doubt reg. the parallelCalls one. As it's more related to the submit logic, rather than to the policy itself.